File: core/semantic/entity_resolution_engine.py
from __future__ import annotations

from typing import Any, Dict, List


def resolve_semantic_entities(
    entities: List[Dict[str, Any]],
) -> Dict[str, Any]:
    canonical: Dict[str, str] = {}
    resolved: List[Dict[str, Any]] = []

    for entity in sorted(entities, key=lambda item: item.get("id", "")):
        label = str(entity.get("label", entity.get("type", ""))).lower().strip()
        canonical_id = canonical.get(label)
        if not canonical_id:
            canonical_id = str(entity.get("id", ""))
            canonical[label] = canonical_id

        resolved.append({
            **entity,
            "canonical_id": canonical_id,
            "resolved": True,
        })

    return {
        "entities": resolved,
        "canonical_map": canonical,
        "bounded": True,
    }

File: src/semantic/entityResolutionEngine.ts
/**
 * Converted from Python: core/semantic/entity_resolution_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function resolveSemanticEntities(entities: any): any {
  var canonical: Record<string, any> = {};
  var resolved: any[] = [];
  var entity: any;
  for (entity of py.iter(py.sorted(entities, {key: ((item: any) => py.get(item, "id", "")) as (item: any) => any}))) {
    var label: any = py.strip(String(py.toStr(py.get(entity, "label", py.get(entity, "type", "")))).toLowerCase());
    var canonical_id: any = py.get(canonical, label);
    if (!py.truthy(canonical_id)) {
      canonical_id = py.toStr(py.get(entity, "id", ""));
      py.setItem(canonical, label, canonical_id);
    }
    py.listAppend(resolved, {...(entity), "canonical_id": canonical_id, "resolved": true});
  }
  return {"entities": resolved, "canonical_map": canonical, "bounded": true};
}
